addons/mail/static/src/messaging/entity/entity_core.js

```diff
@@ -107,49 +71,180 @@ function checkRelations(Entities) {
                     }", should instead refer to this entity or parented entity: ${
                         allSelfAndParentNames.map(name => `"${name}"`).join(', ')
                     }?)`
                 );
             }
             if (
-                (relation.type === 'many2many' && inverseRelation.type !== 'many2many') ||
-                (relation.type === 'one2one' && inverseRelation.type !== 'one2one') ||
-                (relation.type === 'one2many' && inverseRelation.type !== 'many2one') ||
-                (relation.type === 'many2one' && inverseRelation.type !== 'one2many')
+                (field.type === 'many2many' && inverseField.type !== 'many2many') ||
+                (field.type === 'one2one' && inverseField.type !== 'one2one') ||
+                (field.type === 'one2many' && inverseField.type !== 'many2one') ||
+                (field.type === 'many2one' && inverseField.type !== 'one2many')
             ) {
                 throw new Error(
                     `Mismatch relations types "${
                         Entity.name
                     }/${
-                        relationName
+                        fieldName
                     }" (${
-                        relation.type
+                        field.type
                     }) and "${
                         RelatedEntity.name
                     }/${
-                        relation.inverse
+                        field.inverse
                     }" (${
-                        inverseRelation.type
+                        inverseField.type
                     }).`
                 );
             }
         }
     }
 }
 
+/**
+ * @private
+ * @param {string} entityName
+ * @returns {Object}
+ */
+function _getEntryFromEntityName(entityName) {
+    if (!registry[entityName]) {
+        registry[entityName] = {
+            dependencies: [],
+            factory: undefined,
+            name: entityName,
+            patches: [],
+        };
+    }
+    return registry[entityName];
+}
+
+/**
+ * @private
+ * @param {mail.messaging.entity.Entity} Entity class
+ * @param {string} fieldName
+ * @param {Object} field
+ * @return {Array} [inverseFieldName, inverseField]
+ */
+function _inverseRelation(Entity, fieldName, field) {
+    const inverseFieldName = `_inverse_${Entity.name}/${fieldName}`;
+    const relFunc =
+         field.type === 'many2many' ? many2many
+        : field.type === 'many2one' ? one2many
+        : field.type === 'one2many' ? many2one
+        : field.type === 'one2one' ? one2one
+        : undefined;
+    if (!relFunc) {
+        throw new Error(`Cannot compute inverse Relation of "${Entity.name}/${fieldName}".`);
+    }
+    const inverseField = relFunc(Entity.name, { inverse: fieldName });
+    return [inverseFieldName, inverseField];
+}
+
+/**
+ * @private
+ * @param {Object} Entities
+ */
+function _normalizeFields(Entities) {
+    for (const Entity of Object.values(Entities)) {
```

Comment:
Add inline comments to explain the goal of each step, as there are too many lines to understand at first glance.